sql/src/main/java/io/crate/operation/reference/sys/cluster/ClusterSettingsExpression.java

```diff
@@ -21,139 +21,22 @@
 import io.crate.operation.reference.NestedObjectExpression;
-import io.crate.types.DataType;
-import org.apache.logging.log4j.Logger;
 import org.elasticsearch.cluster.service.ClusterService;
 import org.elasticsearch.common.inject.Inject;
-import org.elasticsearch.common.logging.Loggers;
-import org.elasticsearch.common.settings.ClusterSettings;
-import org.elasticsearch.common.settings.Settings;
-
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 
 
 public class ClusterSettingsExpression extends NestedObjectExpression {
 
     public static final String NAME = "settings";
-    private final static Logger LOGGER = Loggers.getLogger(ClusterSettingsExpression.class);
-
-    static class SettingExpression implements ReferenceImplementation<Object> {
-        private final Map<String, Object> values;
-        private final String name;
-        private final DataType dataType;
-
-        SettingExpression(Setting<?, ?> setting, Map<String, Object> values) {
-            this.name = setting.settingName();
-            this.values = values;
-            this.dataType = setting.dataType();
-        }
-
-        @Override
-        public Object value() {
-            return dataType.value(this.values.get(name));
-        }
-    }
-
-    static class NestedSettingExpression extends NestedObjectExpression {
-
-        private final Map<String, Object> values;
-
-        NestedSettingExpression(Setting<?, ?> setting, Map<String, Object> values) {
-            this.values = values;
-            addChildImplementations(setting.children());
-        }
-
-        private void addChildImplementations(List<Setting> childSettings) {
-            for (Setting childSetting : childSettings) {
-                if (childSetting.children().isEmpty()) {
-                    childImplementations.put(childSetting.name(), new SettingExpression(childSetting, values));
-                } else {
-                    childImplementations.put(childSetting.name(), new NestedSettingExpression(childSetting, values));
-                }
-            }
-        }
-    }
-
-    private final ConcurrentHashMap<String, Object> values = new ConcurrentHashMap<>();
-    private final ClusterSettings clusterSettings;
-    private final ClusterService clusterService;
 
     @Inject
-    public ClusterSettingsExpression(Settings settings, ClusterService clusterService) {
-        this.clusterSettings = clusterService.getClusterSettings();
-        this.clusterService = clusterService;
-        setDefaultValues(CrateSettings.SETTINGS);
-        applyInitialSettingsAndRegisterUpdateConsumer(CrateSettings.SETTINGS, settings);
-        addChildImplementations();
-    }
+    public ClusterSettingsExpression(ClusterService clusterService, CrateSettings crateSettings) {
+        // add all exposed crate and elasticsearch settings
+        childImplementations.putAll(crateSettings.referenceImplementationMap());
```

Comment:
Wouldn't it be possible to do the lookups directly on the referenceImplementationMap instead of creating another copy?